spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HealthCheckServiceInstanceListSupplier.java

```diff
@@ -0,0 +1,136 @@
+		implements ServiceInstanceListSupplier {
+
+	private static final Log LOG = LogFactory
+			.getLog(HealthCheckServiceInstanceListSupplier.class);
+
+	private final ServiceInstanceListSupplier delegate;
+
+	private final LoadBalancerProperties.HealthCheck healthCheck;
+
+	private final WebClient webClient;
+
+	private final String defaultHealthCheckPath;
+
+	private List<ServiceInstance> instances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	private List<ServiceInstance> healthyInstances = Collections
+			.synchronizedList(new ArrayList<>());
+
+	public HealthCheckServiceInstanceListSupplier(ServiceInstanceListSupplier delegate,
+			LoadBalancerProperties.HealthCheck healthCheck, WebClient webClient) {
+		this.delegate = delegate;
+		this.healthCheck = healthCheck;
+		defaultHealthCheckPath = healthCheck.getPath().getOrDefault("default",
+				"/actuator/health");
+		this.webClient = webClient;
+		initInstances();
+
+	}
+
+	private void initInstances() {
+		delegate.get().subscribe(delegateInstances -> {
+			instances.clear();
+			instances.addAll(delegateInstances);
+		});
+
+		Flux<List<ServiceInstance>> healthCheckFlux = healthCheckFlux();
+
+		healthCheckFlux.subscribe(verifiedInstances -> {
+			healthyInstances.clear();
+			healthyInstances.addAll(verifiedInstances);
+		});
+	}
+
+	protected Flux<List<ServiceInstance>> healthCheckFlux() {
+		return Flux.create(emitter -> Schedulers
+				.newSingle("Health Check Verifier: " + getServiceId(), true)
+				.schedulePeriodically(() -> {
+					List<ServiceInstance> verifiedInstances = new ArrayList<>();
+					Flux.fromIterable(instances).filterWhen(this::isAlive)
+							.subscribe(serviceInstance -> {
+								verifiedInstances.add(serviceInstance);
+								emitter.next(verifiedInstances);
+							});
+				}, healthCheck.getInitialDelay(), healthCheck.getInterval().toMillis(),
+						TimeUnit.MILLISECONDS),
+				FluxSink.OverflowStrategy.LATEST);
+	}
+
+	@Override
+	public String getServiceId() {
+		return delegate.getServiceId();
+	}
+
+	@Override
+	public Flux<List<ServiceInstance>> get() {
+		if (!healthyInstances.isEmpty()) {
+			return Flux.defer(() -> Flux.fromIterable(healthyInstances).collectList());
+		}
+		// If there are no healthy instances, it might be better to still retry on all of
+		// them
+		if (LOG.isWarnEnabled()) {
+			LOG.warn(
+					"No verified healthy instances were found, returning all listed instances.");
+		}
+		return Flux.defer(() -> Flux.fromIterable(instances).collectList());
+	}
+
+	protected Mono<Boolean> isAlive(ServiceInstance serviceInstance) {
+		String healthCheckPropertyValue = healthCheck.getPath()
+				.get(serviceInstance.getServiceId());
+		String healthCheckPath = healthCheckPropertyValue != null
+				? healthCheckPropertyValue : defaultHealthCheckPath;
+		return webClient.get()
+				.uri(UriComponentsBuilder.fromUri(serviceInstance.getUri())
+						.path(healthCheckPath).build().toUri())
+				.exchange()
```

Comment:
@OlgaMaciaszek  @spencergibb  
this code uses `exchange()` but does not consume `ClientResponse` body  
see https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html#webflux-client-exchange  

> Unlike retrieve(), when using exchange(), it is the responsibility of the application to consume any response content regardless of the scenario (success, error, unexpected data, etc). Not doing so can cause a memory leak. The Javadoc for `ClientResponse lists all the available options for consuming the body. Generally prefer using retrieve() unless you have a good reason for using exchange() which does allow to check the response status and headers before deciding how to or if to consume the response.
--